### PR TITLE
Performance optimization

### PR DESCRIPTION
**Pull request by piaskowyk**
## Benchmarks

<details>
<summary>bezier - before 11 FPS, after 52 FPS</summary>

```js
import Animated, {
  useSharedValue,
  withTiming,
  useAnimatedStyle,
  Easing
} from 'react-native-reanimated';
import { View, Button } from 'react-native';
import React from 'react';

function AnimatedStyleUpdateExample(): React.ReactElement {
  const randomWidth = useSharedValue(10);

  const config = {
    duration: 500,
    easing: Easing.bezier(0.5, 0.01, 0, 1),
  };

  const itemCount = 100;

  const components = [];
  for(let i = 0; i < itemCount; i++) {
    components.push(
      <Animated.View key={i}
        style={[
          { width: 100, height: 5, backgroundColor: 'black', margin: 1 },
          useAnimatedStyle(() => {
            return {
              width: withTiming(randomWidth.value, config),
            };
          }),
        ]}
      />
    )
  }

  return (
    <View
      style={{
        flex: 1,
        flexDirection: 'column',
      }}>
      <Button
        title="toggle"
        onPress={() => { randomWidth.value = Math.random() * 350; }}
      />
      { components }
    </View>
  );
}
export default AnimatedStyleUpdateExample;
```
</details>

<details>
<summary>withTiming rotate - before 35 FPS, after 60 FPS</summary>

```js
import Animated, {
  useSharedValue,
  withTiming,
  useAnimatedStyle,
  withRepeat,
} from 'react-native-reanimated';
import { View, Button } from 'react-native';
import React from 'react';

const dummy = new Array(100).fill(1);
function AnimatedStyleUpdateExample() {
  return (
    <View style={{ flex: 1, flexDirection: "row", flexWrap: "wrap" }}>
      {dummy.map((_, i) => {

        const val = useSharedValue(0);
        const DEG = Math.PI * 2 * 10;
        val.value = withRepeat(
          withTiming(DEG, { duration: 5000 }),
          -1, true
        );

        const style = useAnimatedStyle(() => {
          'worklet'
          return {
            transform: [
               { rotate: val.value + 'deg' },
            ],
          };
        });

        return (
          <Animated.View
            key={i}
            style={[
              { width: 50, height: 50, borderWidth: 1 },
              style,
            ]}
          />
        );
      })}
    </View>
  );
}

export default AnimatedStyleUpdateExample;
```
</details>

<details>
<summary>colors interpolation - before 11 FPS, after 16 FPS</summary>

```js
import Animated, {
  useSharedValue,
  withTiming,
  useAnimatedStyle,
  withRepeat,
  interpolateColor,
} from 'react-native-reanimated';
import { View } from 'react-native';
import React from 'react';

const dummy = new Array(100).fill(1);
function AnimatedStyleUpdateExample() {
  return (
    <View style={{ flex: 1, flexDirection: "row", flexWrap: "wrap" }}>
      {dummy.map((_, i) => {

        const color = useSharedValue(0);
        color.value = withRepeat(
          withTiming(1, { duration: 5000 }),
          -1, true
        );

        const style = useAnimatedStyle(() => {
          'worklet'
          return {
            backgroundColor: interpolateColor(color.value, [0, 1], ['#ffffff', '#000000'])
          };
        });

        return (
          <Animated.View
            key={i}
            style={[
              { width: 50, height: 50, borderWidth: 1 },
              style,
            ]}
          />
        );
      })}
    </View>
  );
}

export default AnimatedStyleUpdateExample;
```
</details>

I tested it on a regular iOS simulator so the optimization results are just indicative.

## Checklist

- [x] Included code example that can be used to test this change
- [x] Updated TS types
- [x] Added TS types tests
- [x] Ensured that CI passes

### CODE REVIEW

**Comment by mrousavy** [comment]
@piaskowyk everytime we run RuntimeDecorator::isXRuntime() we do a property getter, do you think we can optimize that in some way? It would be awesome if we could store a C++ property (enum?) in the jsi::Runtime instance that represents it's type. (Worklet & UI, just Worklet, React-JS)

**Comment by mrousavy** [comment]
Also, what's `styleUpdater()`? Is that the `valueSetter` in the RuntimeManager?

File: src/reanimated2/Colors.ts [hunk 1]
```diff
@@ -656,83 +656,118 @@ export function toRGBA(HSVA: [number, number, number, number]): string {
...
+  return newCache;
+};
+
+interface InterpolateCacheHSV {
+  h: number[];
+  s: number[];
+  v: number[];
+}
+
+const interpolateCacheHSV: { [name: string]: InterpolateCacheHSV } = {};
```
Comment (mrousavy) [inline src/reanimated2/Colors.ts:724]: a `Map` _might_ be faster here

File: src/reanimated2/Colors.ts [hunk 2]
```diff
@@ -656,83 +656,118 @@ export function toRGBA(HSVA: [number, number, number, number]): string {
...
 };
 
+interface InterpolateCacheRGBA {
+  r: number[];
+  g: number[];
+  b: number[];
+  a: number[];
+}
+
+const interpolateCacheRGBA: { [name: string]: InterpolateCacheRGBA } = {};
```
Comment (mrousavy) [inline src/reanimated2/Colors.ts:688]: also `Map` instead of object (it's faster at insertions)

**Comment by mrousavy** [comment]
Also, https://github.com/software-mansion/react-native-reanimated/pull/1863 seems to be somewhat related

**Comment by piaskowyk** [comment]
> @piaskowyk everytime we run RuntimeDecorator::isXRuntime() we do a property getter, do you think we can optimize that in some way? It would be awesome if we could store a C++ property (enum?) in the jsi::Runtime instance that represents it's type. (Worklet & UI, just Worklet, React-JS)

Good point 👍, I did it here: https://github.com/software-mansion/react-native-reanimated/pull/1879/commits/48e1a2ae7c4ab484c95bb7f167e48efaa6babf21

> What's `styleUpdater()`? Is that the `valueSetter` in the RuntimeManager? Couldn't that be written in C++ instead of JS?

The styleUpdater() is JS function calling by mapper: https://github.com/software-mansion/react-native-reanimated/blob/master/src/reanimated2/Hooks.ts#L225
But I speed up this method. 🚀

File: src/reanimated2/Colors.ts [hunk 3]
```diff
@@ -711,7 +714,13 @@ const getInterpolateCacheRGBA = (
     }
   }
   const newCache = { r, g, b, a };
+  const overrideHash = hashOrderRGBA[curentHashIndexRGBA];
```
Comment (Szymon20000) [inline src/reanimated2/Colors.ts:717]: Could you please use LRU Cache here? https://stackoverflow.com/a/46432113

**Review by Szymon20000 — APPROVED** [review]

**Comment by NikitaDudin** [comment]
App crashes on using style transform prop.

Sample code:
```ts
import React, { useEffect } from 'react';
import { StyleSheet, View } from 'react-native';
import Animated, {
  Easing,
  interpolate,
  useAnimatedStyle,
  useSharedValue,
  withRepeat,
  withTiming,
} from 'react-native-reanimated';

const styles = StyleSheet.create({
  main: {
    flex: 1,
    backgroundColor: '#accc00',
    justifyContent: 'center',
    alignItems: 'center',
  },
  box: {
    width: 100,
    height: 100,
    backgroundColor: '#a00',
  },
});

function Comp(): JSX.Element {
  const progress = useSharedValue(0);

  const style = useAnimatedStyle(() => ({
    width: interpolate(progress.value, [0, 0.5, 1], [100, 200, 100]),
    transform: [{ rotate: `${interpolate(progress.value, [0, 1], [0, Math.PI * 2])}rad` }],
  }), [progress]);

  useEffect(() => {
    progress.value = withRepeat(withTiming(1, { duration: 2000, easing: Easing.linear }), -1, false);
  }, [progress]);

  return (
    <View style={styles.main}>
      <Animated.View style={[styles.box, style]} />
    </View>
  );
}

export default Comp;

```

Error message:
```
node_modules/react-native-reanimated/src/reanimated2/Hooks.ts (162:0)

Possible solutions are:
a) If you want to synchronously execute this method, mark it as a Worklet
b) If you want to execute this method on the JS thread, wrap it using runOnJS
```

When comment out the code in isAnimated, then error disappears.
```ts
function isAnimated(prop) {
  'worklet';
  if (Array.isArray(prop)) {
    // comment problem section - START
    // for (const item of prop) {
    //   for (const key in item) {
    //     if (item[key].onFrame !== undefined) {
    //       return true;
    //     }
    //   }
    // }
    // comment problem section - END
    return false;
  }
  return prop.onFrame !== undefined;
}
```

**Pull request merged**